Add Logger to Output Query Logs

meri025 opened the pull request:
### Summary

This pull request introduces a logger to output query logs when the logger is specified in the setup. This enhancement will allow for better tracking and debugging of queries, especially during development and testing.

#### Changes

- Added a logger that can be configured to output query logs.
- Modified the setup to accept a logger configuration.
- Updated README.md to reflect these changes.
- The logger configuration is optional. If not specified, the application will continue to run without logs.

### Other Information

- Corrected a typo in the README.md from "DatabaseRewinder" to the correct term.

### Checklist

* [x] By placing an "x" in the box, I hereby understand, accept and agree to be bound by the terms and conditions of the [Contribution License Agreement](https://dena.github.io/cla/).

Commit:
feature: add logger

Commit:
fixed typo

genya0407 commented on README.md:
````diff
@@ -105,6 +105,17 @@ except_tables = [
 
+### Logging
+
+If you want to enable logging, do the following:
+
+* Generate db_configs from `ActiveRecord::Base.configurations`
+* Pass `logger: Logger` to `MysqlRewinder.setup`.
+
+```ruby
+MysqlRewinder.setup(db_configs, logger: Logger.new(STDOUT))
+```
````
I think we can simplify like:

```suggestion
If you want to enable logging, specify `logger` for `MysqlRewinder.setup`

```ruby
MysqlRewinder.setup(db_configs, logger: Logger.new(STDOUT))
```
```

genya0407 commented on README.md:
```diff
@@ -89,7 +89,7 @@ MysqlRewinder.setup(db_configs, adapter: :mysql2)
 
 * Generate db_configs from `ActiveRecord::Base.configurations`
-* Pass `ActiveRecord::SchemaMigration.new(nil).table_name` and `ActiveRecord::Base.internal_metadata_table_name` to `DatabaseRewinder.setup` as `except_tables`
+* Pass `ActiveRecord::SchemaMigration.new(nil).table_name` and `ActiveRecord::Base.internal_metadata_table_name` to `MysqlRewinder.setup` as `except_tables`
```
Thanks!

genya0407 commented on lib/mysql_rewinder/cleaner.rb:
```diff
@@ -31,5 +35,22 @@ def all_tables
+      res = yield
+      duration = (Time.now - start_ts) * 1000
+
+      name = "Cleaner SQL (#{duration.round(1)}ms)"
```
Add some message like the following, so that users can easily distinguish logs of MysqlRewinder from other query logs.

```suggestion
      name = "[MysqlRewinder] Cleaner SQL (#{duration.round(1)}ms)"
```

genya0407 commented on lib/mysql_rewinder/cleaner.rb:
```diff
@@ -31,5 +35,22 @@ def all_tables
+
+    private
+
+    def log_sql(sql)
```
To simplify code, I think we must rename this method into `log_and_execute(sql)`, and execute the query inside this method (replace `yield` by `@client.execute(sql)`).

meri025 replied:
Thanks. Fixed for https://github.com/DeNA/mysql_rewinder/pull/4/commits/69070ff65426744b12b767cd63df71029bed3bc5

genya0407 commented on lib/mysql_rewinder/cleaner.rb:
```diff
@@ -31,5 +35,22 @@ def all_tables
+    private
+
+    def log_sql(sql)
+      return yield unless @logger.debug?
```
With this, I understand that the logging code below is executed even if `@logger` is NullLogger, that isn't desirable.

Instead, we should set `@logger = nil` if logger is not passed.
Then we can write like:

```ruby
def log_and_execute(sql)
  return @client.execute(sql) unless @logger&.debug?

  ...
end
```

meri025 replied:
Thanks. Fixed for https://github.com/DeNA/mysql_rewinder/pull/4/commits/69070ff65426744b12b767cd63df71029bed3bc5

genya0407 reviewed (commented):
Thank you for contributing!
I have some comments. Please check them.

Commit:
Update README.md

Co-authored-by: Yusuke Sangenya <longinus.eva@gmail.com>

Commit:
Update lib/mysql_rewinder/cleaner.rb

Co-authored-by: Yusuke Sangenya <longinus.eva@gmail.com>

Commit:
Prevent logging logic from executing when using NullLogger

Commit:
Update .rbs

The type definition of the logger faced the same issue as described in https://github.com/ruby/rbs/issues/1482, so it was set to untyped.

genya0407 commented on lib/mysql_rewinder/cleaner.rb:
```diff
@@ -20,8 +21,11 @@ def clean(tables:)
+      delete_sql = target_tables.map { |table| "DELETE FROM #{table}" }.join(';')
+
+      log_and_execute(disable_foreign_key_checks)
+      log_and_execute(delete_sql)
```
These variables are not seemed to be necessary.

```suggestion
      log_and_execute("SET FOREIGN_KEY_CHECKS = 0;")
      log_and_execute(target_tables.map { |table| "DELETE FROM #{table}" }.join(';'))
```

genya0407 commented on sig/mysql_rewinder.rbs:
```diff
@@ -5,13 +5,14 @@ class MysqlRewinder
   # @inserted_table_record_dir: Pathname
   @cleaners: Array[Cleaner]
   @inserted_tables: Set[String]
+  @logger: untyped
```
Maybe we can fix `untyped` to `Logger?`

meri025 replied:
The type definition of the logger faced the same issue as described in https://github.com/ruby/rbs/issues/1482, so it was set to untyped.

genya0407 replied:
I see. This is OK for now.

Commit:
Update lib/mysql_rewinder/cleaner.rb

Co-authored-by: Yusuke Sangenya <longinus.eva@gmail.com>
